packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -285,11 +395,31 @@ export class TokenDetectionController extends StaticIntervalPollingController<
   }
 
   /**
-   * Triggers asset ERC20 token auto detection for each contract address in contract metadata on mainnet.
+   * Restart token detection polling period and call detectNewTokens
+   * in case of address change or user session initialization.
    *
-   * @param options - Options to detect tokens.
+   * @param options - Options for restart token detection.
+   * @param options.selectedAddress - the selectedAddress against which to detect for token balances
+   * @param options.chainId - the chainId against which to detect for token balances
+   */
+  async #restartTokenDetection({
+    selectedAddress,
+    chainId,
+  }: Partial<{ selectedAddress: string; chainId: Hex }> = {}) {
+    await this.detectTokens({
+      accountAddress: selectedAddress,
+      networkClientId: chainId,
+    });
+    this.setIntervalLength(DEFAULT_INTERVAL);
```

Comment:
This is strange. Why would we want to reset the interval when the chain changes? Something to think about for a future PR.

Reply:
> Should `networkDidChange` listener maybe call token detection if `selectedNetworkClientId` is changed instead of chainId? Or maybe it should call `detectTokens` instead of `#restartTokenDetection`?

Yes, I agree with both of these — use `selectedNetworkClientId` instead of `chainId`, and go straight to `detectTokens`.

Again, maybe something we do in a followup PR?

Reply:
I think it makes sense to do this in this PR, since determining where and how to use `#restartTokenDetection` falls under its purview. 

Applied here: c5188899c36a4acce0518229fe0f25c855472fcb